x-pack/filebeat/input/cometd/input.go

```diff
@@ -0,0 +1,196 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package cometd
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/elastic/beats/v7/filebeat/channel"
+	"github.com/elastic/beats/v7/filebeat/input"
+	"github.com/elastic/beats/v7/libbeat/beat"
+	"github.com/elastic/beats/v7/libbeat/common"
+	"github.com/elastic/beats/v7/libbeat/common/atomic"
+	"github.com/elastic/beats/v7/libbeat/logp"
+
+	bay "github.com/elastic/bayeux"
+)
+
+const (
+	inputName = "cometd"
+)
+
+// Run starts the input worker then returns. Only the first invocation
+// will ever start the worker.
+func (in *cometdInput) Run() {
+	var err error
+	in.workerOnce.Do(func() {
+		in.workerWg.Add(1)
+		go func() {
+			in.log.Info("Input worker has started.")
+			defer in.log.Info("Input worker has stopped.")
+			defer in.workerWg.Done()
+			defer in.workerCancel()
+			in.b = bay.Bayeux{}
+			in.creds, err = bay.GetSalesforceCredentials(in.authParams)
+			if err != nil {
+				in.log.Error("not able to get access token", err)
+				return
+			}
+			if err := in.run(); err != nil {
+				in.log.Error(err)
+				return
+			}
+		}()
+	})
+}
+
+func (in *cometdInput) run() error {
+	in.out = in.b.Channel(in.out, "-1", *in.creds, in.config.ChannelName)
+
+	var event Event
+	for e := range in.out {
```

Comment:
This for loop needs to be able to exit when `workerCtx` is cancelled or the goroutine this is run from can never unblock.

You will need to convert it to something like:

```go
for {
  select {
  case <-in.workerCtx.Done():
    return in.workerCtx.Err()
  case e := <- in.out:
    // handle e
  }
}
```

The other option is to make it is so a context is passed to the `bayeux` Channel instance and so it can close the `out` channel as the signal to exit here (via `e, closed := <- in.out`).